src/Xrefcheck/Core.hs

```diff
@@ -17,6 +17,7 @@ import Data.Default (Default (..))
 import qualified Data.List as L
```

Comment:
I gave this another try.

I setup a small test server that does this:
1. The first time it's called, it waits for 3 seconds and then responds with 439 and `Retry-after: 5`
2. The second time it's called, it waits for 3 seconds and then responds with 200.

I've recorded `xrefcheck`'s output:

https://user-images.githubusercontent.com/718166/140096617-992f5029-bfe0-4a26-a568-8d80cdfd97e1.mp4

I noticed a couple of things:
1. The progress bar is not updated after the first call
2. Xrefcheck took a total of 6 seconds to complete. This suggests xrefcheck did not wait for 5 seconds, as instructed by the `Retry-After` header.

Reply:
Does the video capture the behavior of xrefcheck when the server is called the second time or the first time?
If it's called the first time, then, as far as I can understand, the server waited 3 seconds and sent an HTTP 4**3**9 response (don't know if it's a typo or not, but I am going to assume it is); somehow, xrefcheck ignored it even with the `check-localhost` flag present, didn't even try to process the error (which is presumably followed from the fact that the progress bar stayed the same the whole time), didn't classify it as 'fixable', and went on waiting another 3 seconds. 
A very peculiar behavior from xrefcheck... :thinking:

Reply:
Okay, I've set up my test server as well to observe how xrefcheck would behave in precisely the exact same scenario you've set up, and I think I've deduced the reason behind that seemingly strange behavior in your video. Here were the steps I reproduced:
1. I created a simple flask app that would set up and run a local 127.0.0.1 server.
2. In the script there's a global switch that, when it's at 0, would send a 429 response with the Retry-After header being set to 5, and when at 1 would send a 200 response. Each time the response object is constructed (in accordance with the switch value), the switch changes: `switch = 1 - switch`.
3. I print out the value of the switch each time I try to connect to the local server **before** returning the response object.

So, I ran xrefcheck on 127.0.0.1 while the flask app was running in the background, and my behavior **precisely matched** yours, one-to-one. But the confusion faded away when I looked at the Python run tool window:
![image](https://user-images.githubusercontent.com/40356589/140238706-b9d7b0f5-9b10-4658-bf47-25450a24b31b.png)
This matches the definition of `Xrefcheck.Verify.checkExternalResource.checkHttp`!
```
checkHttp :: URI -> ExceptT VerifyError IO ()
checkHttp uri = makeHttpRequest uri HEAD 0.3 `catchError` \_ ->
  makeHttpRequest uri GET 0.7
```
That means the GET request is performed right after the HEAD request with little to no delay. I wanted to confirm this by debugging the flask app, and, indeed, the hypothesis was confirmed. `makeHttpRequest uri GET 0.7` is _not_ lazy, turns out, and I claim that this is exactly what happened in your video.

The last thing I wanted to do was to ensure that making the switch _at least_ quaternary, not binary, would make xrefcheck behave as expected. If the switch is either 0 or 1, send the 429 response; otherwise -- send 200.

https://user-images.githubusercontent.com/40356589/140242118-de07c5e1-9cd3-467b-99e0-c9bfdb474092.mp4

3 seconds: first HEAD request
3 seconds: first GET request
5 seconds: retry-after
3 seconds: second HEAD request
3 seconds: second GET request
summing up to 17 seconds.

Reply:
Yup, you're absolutely right, i tried commenting out the `HEAD` request in `checkHttp` and this time it did wait 11s.

It seems like the algorithm should _not_ attempt a `GET` request if the `HEAD` returned a 429.

Reply:
I think I've revealed the final piece to this puzzle with this new observation:

0. First of all, I have preliminarily prepared the server in accordance with your initial setup.
1. In the `checkHttp` definition I've inserted the debugging element to print the request result 
(`Either VerifyError ()`) right after the request itself has been performed. I snapped a screenshot of xrefcheck's execution.
![image](https://user-images.githubusercontent.com/40356589/140439696-60375e59-e220-4cd7-8507-93c04d55fc11.png)
Turns out, this whole time the reason behind the GET request being performed immediately after the HEAD request was because of the **timeout** as a consequence of the fact that the server waits 3 whole seconds.
2. As usual, I wanted to make sure that decreasing the server sleep time to a certain point would make xrefcheck behave as expected and solve the dilemma once and for all. 
    1. Instead of 3 seconds, I chose 2. 
    2. The GET request is, as per your comment, never sent when we hit the 429 error.

    Here is the grand finale:

https://user-images.githubusercontent.com/40356589/140441391-a426bc2b-b891-499c-8747-dd8023b34587.mp4

2 (sleep) + 5 (retry-after) + 2 (sleep) = 9

So, to sum up: 
1. Xrefcheck's behavior from your video is primarily because of the timeout error, but even if the server sleep time was to be decreased and, incidentally, you were to get a 429 error, the behavior would not change because catching an error immediately triggers the GET request to be performed, which, according to your server configuration, calls the server the second time, receiving a 200 response 3 seconds later.
2. `` `catchError` \_ -> makeHttpRequest uri GET 0.7`` _is_ indeed lazy, as it should be (and has always been).
3. The GET request should never be attempted upon receiving a 429 error.

Reply:
Awesome, great job! :smile:  Marking this as resolved.